### ISSUE TITLE
[M13] Swift MLX training step + optimizer, parity vs mlx_train_step

### ISSUE THREAD

**Issue by travisgalloway** [travisgalloway/monica#195]
Part of #163. Extends M11 from inference-only to a **train-capable** Swift backend, per the pivot to training the M13 ~100M TS code model natively.

## Scope
- Implement forward+backward+optimizer in Swift using mlx-swift's autodiff (`mx` gradient transforms), mirroring `src/model/mlx_train_step.py`'s `make_train_step`/`make_sft_train_step` (gradient accumulation, clipping) for the causal AR objective first.
- Mirror the dynamic fp16 loss-scaling policy (`src/train/loss_scale.py`'s portable policy; Swift does the inf/nan check + skip, matching the Python backend's contract).
- **Parity gate:** loss and grad-norm match `src/model/mlx_train_step.py` within tolerance on a fixed-seed toy batch (mirrors `src/conformance/backend_parity.py`'s intent, ported to a Swift-vs-Python check).

## Non-goals (this issue)
DPO/GRPO Swift step factories (Python remains authoritative for post-training in the near term); gradient checkpointing (add once the base step is validated, mirroring the CUDA/MLX backends' `grad_checkpoint` config).

## Depends on
#166 (model port + logit parity) must land first — this issue reuses that model definition.

## Acceptance
A Swift training script trains the toy config for N steps with a decreasing loss; loss/grad-norm parity vs the Python MLX backend on the same fixed-seed batch.

**Comment by travisgalloway** [comment]
Implemented on `feature/195-swift-train-step` (commit 05a2ae4), opening PR now.

**Landed (REPRODUCE):**
- Forward+backward+optimizer via mlx-swift's own autodiff (`MLXNN.valueAndGrad`), mirroring `make_train_step`/`_accumulate_and_step` — grad accumulation/averaging, hand-rolled grad clipping (deliberately NOT `MLXOptimizers.clipGradNorm`, whose clip-boundary differs from Python's `min(1.0, clip/(norm+1e-6))`), and the dynamic fp16 loss-scale policy + skip control flow (`LossScaler.swift`, pure Swift, no MLX import).
- `accumulateAndStep` factored as the exact same shared accumulate→clip→step tail Python's `_accumulate_and_step` is, so a future SFT masked-CE step is a small delta.
- Parity gate (`monica-train`): loss / grad-norm / full gradient tree / post-step weights vs a new K-step trajectory oracle (`scripts/export_parity_fixture.py --train-steps`) on `toy`/`toy-hybrid`/`toy-moe`, at each fixture's own `train_rtol`/`train_atol` (new `meta.json` keys, `2e-4`/`1e-6` — not the fp32 forward/step gate's `1e-4`/`1e-5`, and not the quantized fixtures' `rtol`/`atol`).
- `--overflow-check`: fp16 overflow → skip branch, scale backoff, bit-identical weights, end to end, no fp16 fixture needed.
- `--train --steps N`: the issue's literal decreasing-loss acceptance criterion.
- CI (`swift-engine`) wired with all four new gates after `monica-parity`.

**Deferred per the plan** (see `.claude/plans/issue-195.md`'s REPRODUCE/DEFER table): `grad_checkpoint`, SFT masked-CE, DPO/GRPO Swift step factories, MoE Loss-Free-Balancing (load counting + `setRouteBias` write path) — proposed as a follow-up issue; optimizer-state save/load → #196 (this issue only reads).

**Verified locally:** `.venv/bin/python -m pytest -q -rs` (1582 passed, 13 pre-existing skips), `scripts/smoke_test.py` (fresh toy split, PASSED), `cd swift && swift build && swift run monica-selfcheck` (OK, tokenizer package untouched — `git diff main -- swift/Package.swift` is empty), `cd swift/engine && swift build` (compiles, including `monica-train`), `monica-train --self-test` (OK — pure Swift, runs even where mlx-swift cannot execute here), the new `tests/test_train_parity_fixture_export.py` (passing).

**CI-only** (mlx-swift cannot execute on this dev host — no compiled `default.metallib`, pre-existing/not this issue's to fix): the `monica-train` fixture gate, `--overflow-check`, and `--train` free-run all execute for the first time in the `swift-engine` CI job — that first run is also where the `train_rtol`/`train_atol` bands get calibrated against real achieved max|d| (the runner prints it on success too).

No tolerance was widened to make a test pass; the global fp32 `rtol=1e-4`/`atol=1e-5` constants are untouched.

**Comment by travisgalloway** [comment]
**`/backlog` run: blocked at merge — PR #293 left open for a design call.**

The port itself is fine. 7 of 8 CI checks are green, **including `swift model parity (macOS, mlx-swift)`** — so the actual Swift-vs-Python training-step parity gate this issue exists to establish *passes*.

The single failing check is `full suite + smoke gate (macOS, MLX)`, on the Python-side staleness guard:

```
tests/test_train_parity_fixture_export.py::test_checked_in_toy_train_fixture_matches_todays_train_step
grad.0.layers.0.conv.weight   max|d| = 1.226e-02   (gate: train_rtol=2e-4 / train_atol=1e-6)
```

**What was verified:**
- The test **passes on local Apple Silicon** (`1 passed in 0.61s`) and **fails on the hosted macOS CI runner** — reproduced across all three CI runs on this PR including its first commit (`05a2ae4`), byte-identical each time.
- Not reproducible locally under mlx 0.31.2 or 0.32.0 (CI's version); local always computes the old checked-in value.
- Neither the test nor `swift/engine/Fixtures/toy/train.safetensors` exists on `main` — both are introduced here, so this is not a pre-existing `main` failure.

**The actual defect is in fixture design, not the port.** A checked-in *gradient* oracle is not reproducible across macOS machines/MLX builds — a ~1% divergence in one conv-backward gradient — so any guard comparing it against a freshly computed value will fail on every machine except the one that generated it. Forward/step *logit* fixtures have not shown this; gradients are more sensitive.

**Three ways forward, all design calls rather than remediation:**
1. Generate `train.safetensors` from inside a GitHub Actions macOS runner, making CI the canonical generator (local runs would then need the same treatment as the guard).
2. Keep the guard but scope it — skip unless running on the generating hardware/MLX build, with an explicit reason.
3. Give the staleness guard its own, deliberately looser band, separate from the `train_rtol`/`train_atol` used for the real Swift-vs-Python comparison.

I did **not** widen `train_rtol`/`train_atol` to make this pass — the plan chose `2e-4/1e-6` specifically because the habitual `1e-5` atol exceeds the gradient magnitudes it is meant to check, and loosening it would have converted a real gate into a vacuous one.

Failing job: https://github.com/travisgalloway/monica/actions/runs/31783545832/job/94714302292
Branch `feature/195-swift-train-step` is preserved (not deleted).
